feat: Implement gRPC Reflection Service

jen20 opened the pull request:
## Motivation

Issue #165 describes the motivations for the gRPC Reflection Service, along with the upstream [definition](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md).

## Solution

This pull request implements a new crate, `tonic-reflection` which implements the gRPC Reflection Service and provides a builder for registering protocol buffer `FileDescriptorProto`s. In the first instance, this relies on danburkert/prost#311, or a similar method, to embed encoded versions of the descriptor in code.

This is an early work in progress, and only implements the `ListServices` and `FileContainingSymbol` aspects of the reflection API, though this is enough to interrogate the shape of services and messages using `grpcurl`. The new example server can be used with `grpcurl` to verify this:

```console
$ grpcurl -plaintext '[::1]:50052' list
grpc.reflection.v1alpha.ServerReflection
helloworld.Greeter

$ grpcurl -plaintext '[::1]:50052' describe helloworld.Greeter
helloworld.Greeter is a service:
// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello ( .helloworld.HelloRequest ) returns ( .helloworld.HelloReply );
}

$ grpcurl -plaintext '[::1]:50052' describe helloworld.HelloRequest                                                                                             
helloworld.HelloRequest is a message:
// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}
```

Currently the implementation is fairly rough - the following work is known as needed:

- [x] Implement remaining request types
- [x] Handle dependencies when processing file descriptors during build
- [x] Write end to end tests
- [x] Remove temporary usages of `thiserror` when all variants are known

~To make this usable, we'll also likely depend on the upstream `prost` pull request being merged, and the `prost-types` crate being generated with descriptor embedding enabled.~

~CI will definitely fail, since it references locally modified crates in the `Cargo.toml` file for now - this PR is currently intended to make progress public for interested parties rather than being in a state to merge!~

LucioFranco commented on tonic-reflection/build.rs:
```diff
@@ -0,0 +1,15 @@
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    const OUT_DIR: &str = "src/proto";
```
might want to add that `println` that allows us to only call build.rs if the file has changed.

LucioFranco commented on tonic-reflection/src/lib.rs:
```diff
@@ -0,0 +1,24 @@
+#![doc(
+    html_logo_url = "https://github.com/hyperium/tonic/raw/master/.github/assets/tonic-docs.png"
+)]
+#![doc(html_root_url = "https://docs.rs/tonic/0.2.0")]
```
```suggestion
#![doc(html_root_url = "https://docs.rs/tonic-reflection/0.1.0")]
```

LucioFranco commented on tonic-reflection/src/server.rs:
```diff
@@ -0,0 +1,323 @@
+use tonic::{Request, Response, Status, Streaming};
+
+/// Represents an error in the construction of a gRPC Reflection Service.
+#[derive(Debug, Error)]
```
we could probably do this by hand tbh, I'd rather not pull thisiserror into the stack.

jen20 replied:
That's one of the points in the PR notes above - `thiserror` is just a temporary convenience thing until I know all the variants to implement it by hand in one go. I'll remove it before converting this into a PR.

LucioFranco replied:
oops sorry! missed that

LucioFranco commented on tonic-reflection/src/server.rs:
```diff
@@ -0,0 +1,323 @@
+    /// the gRPC Reflection Service builder.
+    pub fn register_encoded_file_descriptor(
+        mut self,
+        encoded_file_descriptor: &'static [u8],
```
why does this need to be static? Would it better to accept a `&[u8]` and box it?

jen20 commented:
I've done some update work on this, including redirecting `prost` correctly in order that CI can run before https://github.com/danburkert/prost/pull/326 is merged. Using `FileDescriptorSet` instead of `FileDescriptorProto` solves some of the issues. Still not fully implemented however, so I'm going to leave this as a draft PR.

LucioFranco commented:
@jen20 sorry for the delay, what is the status on this? What are we blocked on?

jen20 commented:
@LucioFranco No worries. To actually release this we're blocked on the `prost` change upstream, but there are also some tests which need writing which I've not yet done.

kinosang commented:
any update here?

jen20 commented:
Hi @kinosang, I'm still waiting for https://github.com/danburkert/prost/pull/326 to land to be able to make much progress in getting this merged here.

jen20 commented:
This is still blocked by danburkert/prost#326 which itself is now blocked by danburkert/prost#397.

sgg commented:
@jen20 I believe that with danburkert/prost#409 (included w/ prost v0.7), this change is now possible 😄.

Do you have the time to drive this? If not, I don't mind picking it up.

jen20 commented:
Hi @sgg! I've rebased this PR and squashed the various changes into logical commits that can land on `master`. Since there is some of your code in here (from #539), I've added you as a co-author in both `Cargo.toml` (for `tonic-reflection` - I assume the addition to `tonic-health` in #539 was a typo!), and as a co-author of the commits to reflect the contributions. Thanks for picking it up!

sgg commented:
@jen20 Sweet! I'll close my PR and we can track the work here.

>  I assume the addition to tonic-health in #539 was a typo!

It most certainly was! 😆

jen20 commented:
Thanks @ssg! Good news, after rebase etc, the original set of tests with the `reflection-server` example are still working:

```
$ grpcurl -plaintext '[::1]:50052' list
grpc.reflection.v1alpha.ServerReflection
helloworld.Greeter

$ grpcurl -plaintext '[::1]:50052' describe helloworld.Greeter
helloworld.Greeter is a service:
// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello ( .helloworld.HelloRequest ) returns ( .helloworld.HelloReply );
}

$ grpcurl -plaintext '[::1]:50052' describe helloworld.HelloRequest
helloworld.HelloRequest is a message:
// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}
```

This should be ready for review shortly.

nfrush commented:
I'm interested to see this make it in :)

LucioFranco commented on tonic-build/src/prost.rs:
```diff
@@ -203,6 +205,13 @@ impl Builder {
 
+    /// Generate a file containing the encoded `prost_types::FileDescriptorSet` for protocol buffers
+    /// modules. This is required for implementing gRPC Server Reflection.
+    pub fn file_descriptor_set_path(mut self, path: impl AsRef<Path>) -> Self {
```
I wonder if we should just make this an option to add the descriptor to the generated proto file as a constant?

jen20 replied:
@LucioFranco Perhaps I'm misunderstanding this - wouldn't `prost` need to support that too?

LucioFranco replied:
If we are going through prost to generate this and it puts it into its own file then yes. I may have misunderstood.
